### ISSUE TITLE
Syntax of combiner

### ISSUE THREAD

**Issue by neurorepro** [nipype/pydra#454]
In the same spirit of #453, I was wondering if it would be still possible to consider a change in the syntax of the combiner, to be more intuitive for developer reading code and not familiar to pydra specificities. Combining outputs in the same way as the standard SQL or pandas `group by` could be great for that.

Currently it may be a bit confusing to know if `combine` will group by the argument, or will group by all other values than the one provided in the arugment. Since `group by` has already a clear definition in database management, it could be great to use it. 

So for a similar example as the [related Pydra tutorial on combine](https://github.com/nipype/pydra-tutorial/blob/master/notebooks/3_intro_functiontask_state.ipynb) but with three variables:
```python
@pydra.mark.task
def add_3var(a, b, c):
    return a + b + c
```
I would believe the following change could give more intuitive code:
* Currently
```python
task_3var = add_3var(name="add_3var", a=[1, 10], b=[100, 1000], c=[10000, 100000])
task_3var.split(["a", "b", "c"])
task_3var.combine(["a", "c"])
task_3var()
task_3var.result()
```
Results
```python
[[Result(output=Output(out=10101), runtime=None, errored=False),
  Result(output=Output(out=100101), runtime=None, errored=False),
  Result(output=Output(out=10110), runtime=None, errored=False),
  Result(output=Output(out=100110), runtime=None, errored=False)],
 [Result(output=Output(out=11001), runtime=None, errored=False),
  Result(output=Output(out=101001), runtime=None, errored=False),
  Result(output=Output(out=11010), runtime=None, errored=False),
  Result(output=Output(out=101010), runtime=None, errored=False)]]
```
* Potential change
```python
task_3var = add_3var(name="add_3var", a=[1, 10], b=[100, 1000], c=[10000, 100000])
task_3var.split(["a", "b", "c"])
task_3var.group_by("b")
task_3var()
task_3var.result() # return a list of results group by b
```
Associated results
```python
[[Result(output=Output(out=10101), runtime=None, errored=False),
  Result(output=Output(out=100101), runtime=None, errored=False),
  Result(output=Output(out=10110), runtime=None, errored=False),
  Result(output=Output(out=100110), runtime=None, errored=False)],
 [Result(output=Output(out=11001), runtime=None, errored=False),
  Result(output=Output(out=101001), runtime=None, errored=False),
  Result(output=Output(out=11010), runtime=None, errored=False),
  Result(output=Output(out=101010), runtime=None, errored=False)]]
```
* Currently
```python
task_3var = add_3var(name="add_3var", a=[1, 10], b=[100, 1000], c=[10000, 100000])
task_3var.split(["a", "b", "c"])
task_3var.combine("b")
task_3var()
task_3var.result()
```
Results
```python
[[Result(output=Output(out=10101), runtime=None, errored=False),
  Result(output=Output(out=11001), runtime=None, errored=False)],
 [Result(output=Output(out=100101), runtime=None, errored=False),
  Result(output=Output(out=101001), runtime=None, errored=False)],
 [Result(output=Output(out=10110), runtime=None, errored=False),
  Result(output=Output(out=11010), runtime=None, errored=False)],
 [Result(output=Output(out=100110), runtime=None, errored=False),
  Result(output=Output(out=101010), runtime=None, errored=False)]]
```
* Potential change
```python
task_3var = add_3var(name="add_3var", a=[1, 10], b=[100, 1000], c=[10000, 100000])
task_3var.split(["a", "b", "c"])
task_3var.group_by(["a", "c"])
task_3var()
task_3var.result() # return a list of results group by a, c
```
Associated esults
```python
[[Result(output=Output(out=10101), runtime=None, errored=False),
  Result(output=Output(out=11001), runtime=None, errored=False)],
 [Result(output=Output(out=100101), runtime=None, errored=False),
  Result(output=Output(out=101001), runtime=None, errored=False)],
 [Result(output=Output(out=10110), runtime=None, errored=False),
  Result(output=Output(out=11010), runtime=None, errored=False)],
 [Result(output=Output(out=100110), runtime=None, errored=False),
  Result(output=Output(out=101010), runtime=None, errored=False)]]
```

What do you think ? (@djarecka or @effigies maybe?)

**Comment by satra** [comment]
Here too the possible uses of combine extend significantly beyond the single variable. In leading up to this formulation, we did consider the various components of group by. In general though even group by varies quite a lot across implementations. Often group by is by value of the variable. 

In pydra, just like split can take several complex forms, group by can also take complex forms, since it literally parallels split. This becomes even more complex when you take nested workflows into account. 

I think the best options is to highlight up front to the user that this is really what sets pydra apart from other such engines, and improve the tutorials for mastery of these concepts through exercises.

**Comment by neurorepro** [comment]
Right, i didn't expect it could have some nested or other complex forms. Seems like everything was already accounted for.

**Comment by neurorepro** [comment]
After thinking of it a bit more, it is not clear to me how `combine` mirrors `split`. Would it be possible to have an example of a complex `combine` expression (with a nested formula mixing inner, outer product)  ? For example assuming a similar example as in #453:
```python
task_4var = op_4var(name="op_4var", a=["a1", "a2", "a3", "a4"], b=["b1", "b2", "b3", "b4"], 
                    c=["c1", "c2"], d=["d1", "d2"])
task_4var.split(("a", ("b", ["c", "d"])))
task_4var.combine(<combine_expression>)
task_4var()
task_4var.result()
```

**Comment by satra** [comment]
Some examples (each one is different):

```python
In [9]: task_4var = op_4var(name="op_4var", a=["a1", "a2"], b=[["b1", "b2"], ["b3", "b4"]],
   ...:                     c=["c1", "c2"], d=["d1", "d2"])
   ...: task_4var.split(["a", ["b", ("c", "d")]], cont_dim={"b": 2})
   ...: task_4var.combine("a")
   ...: task_4var()
   ...: task_4var.result()
Out[9]: 
[[Result(output=Output(out='a1 b1 c1 d1'), runtime=None, errored=False),
  Result(output=Output(out='a2 b1 c1 d1'), runtime=None, errored=False)],
 [Result(output=Output(out='a1 b1 c2 d2'), runtime=None, errored=False),
  Result(output=Output(out='a2 b1 c2 d2'), runtime=None, errored=False)],
 [Result(output=Output(out='a1 b2 c1 d1'), runtime=None, errored=False),
  Result(output=Output(out='a2 b2 c1 d1'), runtime=None, errored=False)],
 [Result(output=Output(out='a1 b2 c2 d2'), runtime=None, errored=False),
  Result(output=Output(out='a2 b2 c2 d2'), runtime=None, errored=False)],
 [Result(output=Output(out='a1 b3 c1 d1'), runtime=None, errored=False),
  Result(output=Output(out='a2 b3 c1 d1'), runtime=None, errored=False)],
 [Result(output=Output(out='a1 b3 c2 d2'), runtime=None, errored=False),
  Result(output=Output(out='a2 b3 c2 d2'), runtime=None, errored=False)],
 [Result(output=Output(out='a1 b4 c1 d1'), runtime=None, errored=False),
  Result(output=Output(out='a2 b4 c1 d1'), runtime=None, errored=False)],
 [Result(output=Output(out='a1 b4 c2 d2'), runtime=None, errored=False),
  Result(output=Output(out='a2 b4 c2 d2'), runtime=None, errored=False)]]

In [10]: task_4var = op_4var(name="op_4var", a=["a1", "a2"], b=[["b1", "b2"], ["b3", "b4"]],
    ...:                     c=["c1", "c2"], d=["d1", "d2"])
    ...: task_4var.split(["a", ["b", ("c", "d")]], cont_dim={"b": 2})
    ...: task_4var.combine("b")
    ...: task_4var()
    ...: task_4var.result()
Out[10]: 
[[Result(output=Output(out='a1 b1 c1 d1'), runtime=None, errored=False),
  Result(output=Output(out='a1 b2 c1 d1'), runtime=None, errored=False),
  Result(output=Output(out='a1 b3 c1 d1'), runtime=None, errored=False),
  Result(output=Output(out='a1 b4 c1 d1'), runtime=None, errored=False)],
 [Result(output=Output(out='a1 b1 c2 d2'), runtime=None, errored=False),
  Result(output=Output(out='a1 b2 c2 d2'), runtime=None, errored=False),
  Result(output=Output(out='a1 b3 c2 d2'), runtime=None, errored=False),
  Result(output=Output(out='a1 b4 c2 d2'), runtime=None, errored=False)],
 [Result(output=Output(out='a2 b1 c1 d1'), runtime=None, errored=False),
  Result(output=Output(out='a2 b2 c1 d1'), runtime=None, errored=False),
  Result(output=Output(out='a2 b3 c1 d1'), runtime=None, errored=False),
  Result(output=Output(out='a2 b4 c1 d1'), runtime=None, errored=False)],
 [Result(output=Output(out='a2 b1 c2 d2'), runtime=None, errored=False),
  Result(output=Output(out='a2 b2 c2 d2'), runtime=None, errored=False),
  Result(output=Output(out='a2 b3 c2 d2'), runtime=None, errored=False),
  Result(output=Output(out='a2 b4 c2 d2'), runtime=None, errored=False)]]

In [11]: task_4var = op_4var(name="op_4var", a=["a1", "a2"], b=[["b1", "b2"], ["b3", "b4"]],
    ...:                     c=["c1", "c2"], d=["d1", "d2"])
    ...: task_4var.split(["a", ["b", ("c", "d")]], cont_dim={"b": 2})
    ...: task_4var.combine("c")
    ...: task_4var()
    ...: task_4var.result()
Out[11]: 
[[Result(output=Output(out='a1 b1 c1 d1'), runtime=None, errored=False),
  Result(output=Output(out='a1 b1 c2 d2'), runtime=None, errored=False)],
 [Result(output=Output(out='a1 b2 c1 d1'), runtime=None, errored=False),
  Result(output=Output(out='a1 b2 c2 d2'), runtime=None, errored=False)],
 [Result(output=Output(out='a1 b3 c1 d1'), runtime=None, errored=False),
  Result(output=Output(out='a1 b3 c2 d2'), runtime=None, errored=False)],
 [Result(output=Output(out='a1 b4 c1 d1'), runtime=None, errored=False),
  Result(output=Output(out='a1 b4 c2 d2'), runtime=None, errored=False)],
 [Result(output=Output(out='a2 b1 c1 d1'), runtime=None, errored=False),
  Result(output=Output(out='a2 b1 c2 d2'), runtime=None, errored=False)],
 [Result(output=Output(out='a2 b2 c1 d1'), runtime=None, errored=False),
  Result(output=Output(out='a2 b2 c2 d2'), runtime=None, errored=False)],
 [Result(output=Output(out='a2 b3 c1 d1'), runtime=None, errored=False),
  Result(output=Output(out='a2 b3 c2 d2'), runtime=None, errored=False)],
 [Result(output=Output(out='a2 b4 c1 d1'), runtime=None, errored=False),
  Result(output=Output(out='a2 b4 c2 d2'), runtime=None, errored=False)]]

In [12]: task_4var = op_4var(name="op_4var", a=["a1", "a2"], b=[["b1", "b2"], ["b3", "b4"]],
    ...:                     c=["c1", "c2"], d=["d1", "d2"])
    ...: task_4var.split(["a", ["b", ("c", "d")]], cont_dim={"b": 2})
    ...: task_4var.combine(["b", "c"])
    ...: task_4var()
    ...: task_4var.result()
Out[12]: 
[[Result(output=Output(out='a1 b1 c1 d1'), runtime=None, errored=False),
  Result(output=Output(out='a1 b1 c2 d2'), runtime=None, errored=False),
  Result(output=Output(out='a1 b2 c1 d1'), runtime=None, errored=False),
  Result(output=Output(out='a1 b2 c2 d2'), runtime=None, errored=False),
  Result(output=Output(out='a1 b3 c1 d1'), runtime=None, errored=False),
  Result(output=Output(out='a1 b3 c2 d2'), runtime=None, errored=False),
  Result(output=Output(out='a1 b4 c1 d1'), runtime=None, errored=False),
  Result(output=Output(out='a1 b4 c2 d2'), runtime=None, errored=False)],
 [Result(output=Output(out='a2 b1 c1 d1'), runtime=None, errored=False),
  Result(output=Output(out='a2 b1 c2 d2'), runtime=None, errored=False),
  Result(output=Output(out='a2 b2 c1 d1'), runtime=None, errored=False),
  Result(output=Output(out='a2 b2 c2 d2'), runtime=None, errored=False),
  Result(output=Output(out='a2 b3 c1 d1'), runtime=None, errored=False),
  Result(output=Output(out='a2 b3 c2 d2'), runtime=None, errored=False),
  Result(output=Output(out='a2 b4 c1 d1'), runtime=None, errored=False),
  Result(output=Output(out='a2 b4 c2 d2'), runtime=None, errored=False)]]

In [15]: task_4var = op_4var(name="op_4var", a=["a1", "a2"], b=[["b1", "b2"], ["b3", "b4"]],
    ...:                     c=["c1", "c2"], d=["d1", "d2"])
    ...: task_4var.split(["a", ["b", ("c", "d")]], cont_dim={"b": 2})
    ...: task_4var.combine(["a", "b", "c"])
    ...: task_4var()
    ...: task_4var.result()
Out[15]: 
[Result(output=Output(out='a1 b1 c1 d1'), runtime=None, errored=False),
 Result(output=Output(out='a1 b1 c2 d2'), runtime=None, errored=False),
 Result(output=Output(out='a1 b2 c1 d1'), runtime=None, errored=False),
 Result(output=Output(out='a1 b2 c2 d2'), runtime=None, errored=False),
 Result(output=Output(out='a1 b3 c1 d1'), runtime=None, errored=False),
 Result(output=Output(out='a1 b3 c2 d2'), runtime=None, errored=False),
 Result(output=Output(out='a1 b4 c1 d1'), runtime=None, errored=False),
 Result(output=Output(out='a1 b4 c2 d2'), runtime=None, errored=False),
 Result(output=Output(out='a2 b1 c1 d1'), runtime=None, errored=False),
 Result(output=Output(out='a2 b1 c2 d2'), runtime=None, errored=False),
 Result(output=Output(out='a2 b2 c1 d1'), runtime=None, errored=False),
 Result(output=Output(out='a2 b2 c2 d2'), runtime=None, errored=False),
 Result(output=Output(out='a2 b3 c1 d1'), runtime=None, errored=False),
 Result(output=Output(out='a2 b3 c2 d2'), runtime=None, errored=False),
 Result(output=Output(out='a2 b4 c1 d1'), runtime=None, errored=False),
 Result(output=Output(out='a2 b4 c2 d2'), runtime=None, errored=False)]
```

**Comment by satra** [comment]
just a note that when using `()` any of the variables inside the `()` should work.

**Comment by NIDS2020-instructor** [comment]
Thanks a lot Satra. I see that in this case, a `group_by` could have made sense for the first two examples: 
- `.combine("a")` -->  `.group_by(["b", "c", "d"])`
- `.combine("b")` -->  `.group_by(["a", "c", "d"])`
But because of the dot product `(c,d)`, `.combine("c")` has a different effect than .group_by(["a", "b", "d"])` which would have the result
```python
[[Result(output=Output(out='a1 b1 c1 d1'), runtime=None, errored=False)],
 [Result(output=Output(out='a1 b1 c2 d2'), runtime=None, errored=False)],
 [Result(output=Output(out='a1 b2 c1 d1'), runtime=None, errored=False)],
 [Result(output=Output(out='a1 b2 c2 d2'), runtime=None, errored=False)],
 [Result(output=Output(out='a1 b3 c1 d1'), runtime=None, errored=False)],
 [Result(output=Output(out='a1 b3 c2 d2'), runtime=None, errored=False)],
 [Result(output=Output(out='a1 b4 c1 d1'), runtime=None, errored=False)],
 [Result(output=Output(out='a1 b4 c2 d2'), runtime=None, errored=False)],
 [Result(output=Output(out='a2 b1 c1 d1'), runtime=None, errored=False)],
 [Result(output=Output(out='a2 b1 c2 d2'), runtime=None, errored=False)],
 [Result(output=Output(out='a2 b2 c1 d1'), runtime=None, errored=False)],
 [Result(output=Output(out='a2 b2 c2 d2'), runtime=None, errored=False)],
 [Result(output=Output(out='a2 b3 c1 d1'), runtime=None, errored=False)],
 [Result(output=Output(out='a2 b3 c2 d2'), runtime=None, errored=False)],
 [Result(output=Output(out='a2 b4 c1 d1'), runtime=None, errored=False)],
 [Result(output=Output(out='a2 b4 c2 d2'), runtime=None, errored=False)]]
```
In that case the user would have to mentally parse himself the initial `split` to come up with the correct `group_by` equivalent of  `.combine("c")`, which would be  `.group_by(["a", "b"])`.

Initially i thought reusing a standard `group_by` approach used in `SQL`, `pandas`, etc. would mean less mental gymnastics for the user to use `pydra`, but i can see now the advantages of a split-aware `combine`.

Many thanks for the examples.

**Comment by neurorepro** [comment]
Thanks a lot Satra. I see that in this case, a `group_by` could have made sense for the first two examples: 
- `.combine("a")` -->  `.group_by(["b", "c", "d"])`
- `.combine("b")` -->  `.group_by(["a", "c", "d"])`
But because of the dot product `(c,d)`, `.combine("c")` has a different effect than `.group_by(["a", "b", "d"])` which would have the result
```python
[[Result(output=Output(out='a1 b1 c1 d1'), runtime=None, errored=False)],
 [Result(output=Output(out='a1 b1 c2 d2'), runtime=None, errored=False)],
 [Result(output=Output(out='a1 b2 c1 d1'), runtime=None, errored=False)],
 [Result(output=Output(out='a1 b2 c2 d2'), runtime=None, errored=False)],
 [Result(output=Output(out='a1 b3 c1 d1'), runtime=None, errored=False)],
 [Result(output=Output(out='a1 b3 c2 d2'), runtime=None, errored=False)],
 [Result(output=Output(out='a1 b4 c1 d1'), runtime=None, errored=False)],
 [Result(output=Output(out='a1 b4 c2 d2'), runtime=None, errored=False)],
 [Result(output=Output(out='a2 b1 c1 d1'), runtime=None, errored=False)],
 [Result(output=Output(out='a2 b1 c2 d2'), runtime=None, errored=False)],
 [Result(output=Output(out='a2 b2 c1 d1'), runtime=None, errored=False)],
 [Result(output=Output(out='a2 b2 c2 d2'), runtime=None, errored=False)],
 [Result(output=Output(out='a2 b3 c1 d1'), runtime=None, errored=False)],
 [Result(output=Output(out='a2 b3 c2 d2'), runtime=None, errored=False)],
 [Result(output=Output(out='a2 b4 c1 d1'), runtime=None, errored=False)],
 [Result(output=Output(out='a2 b4 c2 d2'), runtime=None, errored=False)]]
```
In that case the user would have to mentally parse himself the initial `split` to come up with the correct `group_by` equivalent of  `.combine("c")`, which would be  `.group_by(["a", "b"])`.

Initially i thought reusing a standard `group_by` approach used in `SQL`, `pandas`, etc. would mean less mental gymnastics for the user to use `pydra`, but i can see now the advantages of a split-aware `combine`.

Many thanks for the examples.

**Comment by satra** [comment]
a few more examples that will work once #458 is merged

```python
In [3]: task_4var = op_4var(name="op_4var", a=["a1", "a2"], b=[["b1", "b2"], ["b3", "b4"]],
   ...:                     c=["c1", "c2"], d=["d1", "d2"])
   ...: task_4var.split(["a", ("b", ["c", "d"])], cont_dim={"b": 2})
   ...: task_4var.combine(["a"])
   ...: task_4var()
   ...: task_4var.result()
Out[3]: 
[[Result(output=Output(out='a1 b1 c1 d1'), runtime=None, errored=False),
  Result(output=Output(out='a2 b1 c1 d1'), runtime=None, errored=False)],
 [Result(output=Output(out='a1 b2 c1 d2'), runtime=None, errored=False),
  Result(output=Output(out='a2 b2 c1 d2'), runtime=None, errored=False)],
 [Result(output=Output(out='a1 b3 c2 d1'), runtime=None, errored=False),
  Result(output=Output(out='a2 b3 c2 d1'), runtime=None, errored=False)],
 [Result(output=Output(out='a1 b4 c2 d2'), runtime=None, errored=False),
  Result(output=Output(out='a2 b4 c2 d2'), runtime=None, errored=False)]]

In [4]: task_4var = op_4var(name="op_4var", a=["a1", "a2"], b=[["b1", "b2"], ["b3", "b4"]],
   ...:                     c=["c1", "c2"], d=["d1", "d2"])
   ...: task_4var.split(["a", ("b", ["c", "d"])], cont_dim={"b": 2})
   ...: task_4var.combine(["a", "b"])
   ...: task_4var()
   ...: task_4var.result()
Out[4]: 
[Result(output=Output(out='a1 b1 c1 d1'), runtime=None, errored=False),
 Result(output=Output(out='a1 b2 c1 d2'), runtime=None, errored=False),
 Result(output=Output(out='a1 b3 c2 d1'), runtime=None, errored=False),
 Result(output=Output(out='a1 b4 c2 d2'), runtime=None, errored=False),
 Result(output=Output(out='a2 b1 c1 d1'), runtime=None, errored=False),
 Result(output=Output(out='a2 b2 c1 d2'), runtime=None, errored=False),
 Result(output=Output(out='a2 b3 c2 d1'), runtime=None, errored=False),
 Result(output=Output(out='a2 b4 c2 d2'), runtime=None, errored=False)]

In [5]: task_4var = op_4var(name="op_4var", a=["a1", "a2"], b=[["b1", "b2"], ["b3", "b4"]],
   ...:                     c=["c1", "c2"], d=["d1", "d2"])
   ...: task_4var.split(["a", ("b", ["c", "d"])], cont_dim={"b": 2})
   ...: task_4var.combine(["b"])
   ...: task_4var()
   ...: task_4var.result()
Out[5]: 
[[Result(output=Output(out='a1 b1 c1 d1'), runtime=None, errored=False),
  Result(output=Output(out='a1 b2 c1 d2'), runtime=None, errored=False),
  Result(output=Output(out='a1 b3 c2 d1'), runtime=None, errored=False),
  Result(output=Output(out='a1 b4 c2 d2'), runtime=None, errored=False)],
 [Result(output=Output(out='a2 b1 c1 d1'), runtime=None, errored=False),
  Result(output=Output(out='a2 b2 c1 d2'), runtime=None, errored=False),
  Result(output=Output(out='a2 b3 c2 d1'), runtime=None, errored=False),
  Result(output=Output(out='a2 b4 c2 d2'), runtime=None, errored=False)]]
```

**Comment by djarecka** [comment]
@neurorepro - we were going back and forth between `combine` and `groupBy`, and at some point we decided to use the word `combine` and NOT follow the `groupBy` logic. Hopefully by using different word we will not confuse everyone. 
And as you observed, because of the nature of scalar product, some input fields are automatically combined even if they are not part of the `combiner`.  In your example, `combine("c")`, `combine("d")` and `combine("c", "d")` are equivalent.
